File: knowledge3d/core/rpn.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Union


Token = Union[float, int, str]


@dataclass
class RPN:
    max_depth: int = 64
    keep_top: int = 5

    def __post_init__(self) -> None:
        self.keep_top = max(5, self.keep_top)
        self._stack: List[float] = []

    def reset(self) -> None:
        self._stack.clear()

    def push(self, x: float) -> None:
        self._stack.append(float(x))
        self._auto_clean()

    def pop(self) -> float:
        if not self._stack:
            raise RuntimeError("RPN underflow")
        return self._stack.pop()

    def peek(self, n: int = 0) -> float:
        idx = len(self._stack) - 1 - n
        if idx < 0:
            raise RuntimeError("RPN underflow")
        return self._stack[idx]

    def _auto_clean(self) -> None:
        if len(self._stack) > self.max_depth:
            keep = max(1, self.keep_top)
            del self._stack[: len(self._stack) - keep]

    def eval(self, tokens: Iterable[Token]) -> float:
        for t in tokens:
            if isinstance(t, (int, float)):
                self.push(float(t))
                continue
            if t == '+':
                self.push(self.pop() + self.pop())
            elif t == '-':
                b, a = self.pop(), self.pop()
                self.push(a - b)
            elif t == '*':
                self.push(self.pop() * self.pop())
            elif t == '/':
                b, a = self.pop(), self.pop()
                self.push(a / b if b != 0 else 0.0)
            elif t == 'sqrt':
                import math
                self.push(math.sqrt(self.pop()))
            elif t == 'abs':
                self.push(abs(self.pop()))
            elif t == 'dup':
                self.push(self.peek())
            elif t == 'drop':
                self.pop()
            elif t == 'swap':
                a, b = self.pop(), self.pop()
                self.push(a)
                self.push(b)
            else:
                raise ValueError(f"unknown op: {t}")
        return self.peek(0)

    def dot(self, a: Iterable[float], b: Iterable[float]) -> float:
        acc = 0.0
        for x, y in zip(a, b):
            acc = self.eval([acc, x, y, '*', '+'])
        return acc

    def norm(self, a: Iterable[float]) -> float:
        acc = 0.0
        for x in a:
            acc = self.eval([acc, x, x, '*', '+'])
        import math
        return math.sqrt(acc)

    def cosine(self, a: Iterable[float], b: Iterable[float]) -> float:
        d = self.dot(a, b)
        na = self.norm(a)
        nb = self.norm(b)
        if na == 0 or nb == 0:
            return 0.0
        return self.eval([d, na, '/', nb, '/'])

File: viewer/src/rpn.ts
export type RPNToken = number | string;

export interface RPNOptions {
  maxDepth?: number; // total capacity
  keepTop?: number;  // keep at least this many elements after auto-clean
}

export class RPN {
  private stack: number[] = [];
  private maxDepth: number;
  private keepTop: number;

  constructor(opts: RPNOptions = {}) {
    this.maxDepth = opts.maxDepth ?? 64;
    this.keepTop = Math.max(5, opts.keepTop ?? 5); // ensure at least 5 levels of memory
  }

  reset() {
    this.stack.length = 0;
  }

  push(x: number) {
    this.stack.push(x);
    this.autoClean();
  }

  pop(): number {
    const v = this.stack.pop();
    if (v === undefined) throw new Error('RPN underflow');
    return v;
  }

  peek(n = 0): number {
    const idx = this.stack.length - 1 - n;
    if (idx < 0) throw new Error('RPN underflow');
    return this.stack[idx];
  }

  size() {
    return this.stack.length;
  }

  private autoClean() {
    if (this.stack.length > this.maxDepth) {
      // Drop oldest but keep at least keepTop values
      const keep = Math.max(this.keepTop, 1);
      this.stack.splice(0, this.stack.length - keep);
    }
  }

  eval(tokens: RPNToken[]): number {
    for (const t of tokens) {
      if (typeof t === 'number') {
        this.push(t);
        continue;
      }
      switch (t) {
        case '+': this.push(this.pop() + this.pop()); break;
        case '-': {
          const b = this.pop(); const a = this.pop(); this.push(a - b); break;
        }
        case '*': this.push(this.pop() * this.pop()); break;
        case '/': {
          const b = this.pop(); const a = this.pop(); this.push(a / b || 0); break;
        }
        case 'sqrt': this.push(Math.sqrt(this.pop())); break;
        case 'abs': this.push(Math.abs(this.pop())); break;
        case 'dup': this.push(this.peek()); break;
        case 'drop': this.pop(); break;
        case 'swap': { const a = this.pop(), b = this.pop(); this.push(a); this.push(b); break; }
        default:
          throw new Error(`RPN unknown op: ${t}`);
      }
    }
    return this.peek(0);
  }

  dot(a: number[], b: number[]): number {
    const n = Math.min(a.length, b.length);
    let acc = 0;
    for (let i = 0; i < n; i++) {
      acc = this.eval([acc, a[i], b[i], '*', '+']);
    }
    return acc;
  }

  norm(a: number[]): number {
    let acc = 0;
    for (let i = 0; i < a.length; i++) {
      acc = this.eval([acc, a[i], a[i], '*', '+']);
    }
    return Math.sqrt(acc);
  }

  cosine(a: number[], b: number[]): number {
    const d = this.dot(a, b);
    const na = this.norm(a);
    const nb = this.norm(b);
    if (na === 0 || nb === 0) return 0;
    return this.eval([d, na, '/', nb, '/']);
  }
}

